app/src/main/java/com/kks/nimblesurveyjetpackcompose/util/interceptors/CustomAccessTokenInterceptor.kt

```diff
@@ -0,0 +1,64 @@
+package com.kks.nimblesurveyjetpackcompose.util.interceptors
+
+import com.kks.nimblesurveyjetpackcompose.repo.token.TokenRepoImpl
+import com.kks.nimblesurveyjetpackcompose.util.PREF_ACCESS_TOKEN
+import com.kks.nimblesurveyjetpackcompose.util.PREF_REFRESH_TOKEN
+import com.kks.nimblesurveyjetpackcompose.util.PreferenceManager
+import dagger.Lazy
+import kotlinx.coroutines.CoroutineScope
+import kotlinx.coroutines.Dispatchers
+import kotlinx.coroutines.flow.collectLatest
+import kotlinx.coroutines.launch
+import okhttp3.Interceptor
+import okhttp3.Request
+import okhttp3.Response
+import java.net.HttpURLConnection
+
+class CustomAccessTokenInterceptor(
+    private val preferenceManager: PreferenceManager,
+    private val tokenRepo: Lazy<TokenRepoImpl>
+) : Interceptor {
+
+    private var isRefreshed = false
+
+    override fun intercept(chain: Interceptor.Chain): Response {
+
+        val accessToken: String =
+            preferenceManager.getStringData(PREF_ACCESS_TOKEN) ?: ""
+        val request: Request = newRequestWithAccessToken(chain.request(), accessToken)
+
+        val response: Response = chain.proceed(request)
+
+        if (response.code == HttpURLConnection.HTTP_UNAUTHORIZED) {
+            val job = CoroutineScope(Dispatchers.IO).launch {
+                tokenRepo.get().refreshToken(
+                    preferenceManager.getStringData(PREF_REFRESH_TOKEN) ?: ""
+                ).collectLatest {
+                    isRefreshed = true
+                }
+            }
+
+            synchronized(this) {
+                while (!isRefreshed) {
+                    // Wait for cache
+                }
+                isRefreshed = false
+                val newAccessToken: String =
+                    preferenceManager.getStringData(PREF_ACCESS_TOKEN) ?: ""
+                job.cancel()
+                // Access token is refreshed in another thread.
+                if (accessToken != newAccessToken) {
+                    return chain.proceed(newRequestWithAccessToken(request, newAccessToken))
+                }
+            }
+        }
```

Comment:
If I'm not wrong, we don't need this block here. If the response code is `HTTP_UNAUTHORIZED`, it should be taken care of by the `TokenAuthenticator`. No? 🤔 

```suggestion
```

Also, how about renaming this class to `AccessTokenInterceptor`? 💡